### PR TITLE
fix: promtail; clean up metrics generated from logs after a config reload.

### PR DESCRIPTION
**Pull request by ptodev**
It is possible to generate metrics from log lines in Promtail. If the config file is reloaded, those metrics are currently not cleaned up. This causes a few issues:
* If a metric is not necessary anymore, it is still visible on the `/metrics` endpoint.
* A metric with the same name might have a different meaning after the config file reload, so it'd make sense to clean these up.

This will help us fix [a bug](https://github.com/grafana/agent/issues/2754) reported in the Grafana Agent.

I suppose no changelog entry is required because PromQL is already designed to handle counter resets? But if you think it's appropriate, I could add a changelog entry?

I tested this change locally using a config file like this:

<details>
  <summary>Promtail config</summary>

```yaml
server:
  http_listen_port: 9080
  grpc_listen_port: 0
  enable_runtime_reload: true
positions:
  filename: /Users/paulintodev/Desktop/log_metrics_bug/tmp_promtail/positions
clients:
  - url: "https://logs-prod-008.grafana.net/loki/api/v1/push"
    basic_auth:
      username: ""
      password: ""
scrape_configs:
- job_name: log files
  static_configs:
  - labels:
      __path__: /Users/paulintodev/Desktop/log_metrics_bug/*.log
  pipeline_stages:
    - metrics:
        log_lines_total:
          type: Counter
          description: "total number of log lines"
          prefix: my_promtail_custom_
          max_idle_duration: 24h
          config:
            match_all: true
            action: inc
        log_lines_total2:
          type: Counter
          description: "total number of log lines"
          prefix: my_promtail_custom2_
          max_idle_duration: 24h
          config:
            match_all: true
            action: inc
```

</details>

Then I'd do a `curl localhost:9080/reload` and check `http://localhost:9080/metrics`. 

I didn't add a whole lot of unit tests because TBH we should probably focus on adding tests in the Agent, given that we're sunsetting Promtail over time.

### CODE REVIEW

File: clients/pkg/logentry/stages/metrics_test.go [hunk 1]
```diff
@@ -127,6 +127,13 @@ func TestMetricsPipeline(t *testing.T) {
 		strings.NewReader(expectedMetrics)); err != nil {
 		t.Fatalf("mismatch metrics: %v", err)
 	}
+
+	pl.Cleanup()
+
+	if err := testutil.GatherAndCompare(registry,
+		strings.NewReader("")); err != nil {
+		t.Fatalf("mismatch metrics: %v", err)
+	}
```
Comment (cstyan) [inline clients/pkg/logentry/stages/metrics_test.go:136]: shouldn't we specifically check for the absence of the metric?

**Review by cstyan — COMMENTED** [review]
@ptodev need to rebase/merge in master to get some CI updates

File: clients/pkg/logentry/stages/metrics.go [hunk 2]
```diff
@@ -142,6 +142,19 @@ type metricStage struct {
...
+	go func() {
+		defer close(out)
+
+		for e := range in {
+			m.Process(e.Labels, e.Extracted, &e.Timestamp, &e.Line)
+			out <- e
+		}
+	}()
+	return out
+}
```
Comment (cstyan) [inline clients/pkg/logentry/stages/metrics.go:156]: why was this change necessary? just to mimic what `toStage` was giving us before?

File: clients/pkg/logentry/stages/metrics_test.go (hunk 1 above)
Comment (ptodev) [reply clients/pkg/logentry/stages/metrics_test.go:136]: After `pl.Cleanup()`, no metrics should be visible. I don't see an advantage in checking individual metrics when there shouldn't be any of them.

File: clients/pkg/logentry/stages/metrics.go (hunk 2 above)
Comment (ptodev) [reply clients/pkg/logentry/stages/metrics.go:156]: Yes. The metrics stage needs a cleanup function, and `toStage` doesn't add one. Maybe we could avoid the repetition by having another function called `toStageWithCleanup`?

Comment (cstyan) [reply clients/pkg/logentry/stages/metrics.go:156]: I think this way is fine. Even with `toStageWithCleanup` we would still require the addition of `Cleanup` noop functions on all stages except the metric stage.

File: clients/pkg/logentry/stages/metrics_test.go (hunk 1 above)
Comment (cstyan) [reply clients/pkg/logentry/stages/metrics_test.go:136]: what if there's originally two metrics stages, and after we reload there should be only one? we should still check for it's existence and correct value, I think right now we're deleting metrics for stages that will still exist after the reload which would not be correct

Comment (ptodev) [reply clients/pkg/logentry/stages/metrics_test.go:136]: I'd argue that deleting all the metrics is the right thing to do. The fact that a metric exists doesn't mean that it's used in the same way in the new config. It's possible that the user changed the meaning of each metric, but retained the metric names. In that case it would make sense to start with fresh metrics.

In the future we could make the code smart and reload only the stages which changed, but I think that's a separate issue. To do that, we could avoid calling the `Cleanup` method for stages which are completely unchanged as a whole.

Comment (cstyan) [reply clients/pkg/logentry/stages/metrics_test.go:136]: we should document this clearly in the metrics config section https://github.com/grafana/loki/blob/a00f1f1b0b8f536f2cdac2f8857eb40c716aa696/docs/sources/send-data/promtail/configuration.md?plain=1#L678 and maybe as a note here as well

**Review by cstyan — COMMENTED** [review]
@ptodev sorry for the delay, I think we just need some docs, also have you guys been using this code downstream in Agent/Alloy already or would merging this be the first usage of this code?

**Review by cstyan — APPROVED** [review]
I'm going to approve/merge this with a note that if we get reports of issues I'll revert this change. Hopefully we here from the Alloy team after a short period of time whether they have confirmed usages that are working well or not.
